src/graph.c

```diff
@@ -0,0 +1,392 @@
+#include <packetgraph/packetgraph.h>
+#include "brick-int.h"
+#include "graph-int.h"
+
+struct pg_graph *pg_graph_new(const char *name, struct pg_brick *explore)
+{
+	struct pg_error *error = NULL;
+	struct pg_graph *ret = g_new(struct pg_graph, 1);
+
+	ret->name = g_strdup(name);
+	ret->all = g_hash_table_new(g_str_hash, g_str_equal);
+	ret->pollable = NULL;
+
+	if (explore) {
+		pg_graph_explore_ptr(ret, explore, &error);
+		if (error)
+			pg_error_free(error);
+	}
+
+	return ret;
+}
+
+const char *pg_graph_name(struct pg_graph *graph)
+{
+	return graph->name;
+}
+
+void pg_graph_destroy(struct pg_graph *graph)
+{
+	GHashTableIter iter;
+	gpointer key, value;
+
+	g_free(graph->name);
+	g_hash_table_iter_init(&iter, graph->all);
+	while (g_hash_table_iter_next(&iter, &key, &value))
+		pg_brick_destroy(value);
+	g_hash_table_destroy(graph->all);
+	g_slist_free(graph->pollable);
+	g_free(graph);
+}
+
+int pg_graph_explore(struct pg_graph *graph,
+		     const char *brick_name,
+		     struct pg_error **error)
+{
+	struct pg_brick *b = NULL;
+
+	if (brick_name) {
+		b = pg_graph_get(graph, brick_name);
+		if (!b) {
+			*error = pg_error_new("brick %s not found", brick_name);
+			return -1;
+		}
+	}
+	return pg_graph_explore_ptr(graph, b, error);
+}
+
+static inline GList *explore_add(GList *todo,
+				 GList *done,
+				 struct pg_brick *brick)
+{
+	if (brick && !g_list_find(todo, brick) && !g_list_find(done, brick))
+		return g_list_prepend(todo, brick);
+	return todo;
+}
+
+static inline GList *explore_add_neighbours(GList *todo,
+					    GList *done,
+					    struct pg_brick *b)
+{
+	if (b->type == PG_MONOPOLE)
+		return explore_add(todo, done, b->side.edge.link);
+	for (int i = 0; i < MAX_SIDE; i++) {
+		if (b->type == PG_DIPOLE) {
+			todo = explore_add(todo, done, b->sides[i].edge.link);
+		} else {
+			for (int j = 0; j < b->sides[i].max; j++)
+				todo = explore_add(todo, done,
+						   b->sides[i].edges[j].link);
+		}
+	}
+	return todo;
+}
+
+static inline struct pg_brick *get_any_brick(struct pg_graph *graph)
+{
+	struct pg_brick *brick = NULL;
+
+	if (graph->pollable) {
+		brick = graph->pollable->data;
+	} else if (g_hash_table_size(graph->all)) {
+		GHashTableIter iter;
+		gpointer key;
+
+		g_hash_table_iter_init(&iter, graph->all);
+		g_hash_table_iter_next(&iter, &key, (void **)&brick);
+	}
+	return brick;
+}
+
+static inline void empty_graph(struct pg_graph *graph)
+{
+	g_hash_table_steal_all(graph->all);
+	g_slist_free(graph->pollable);
+	graph->pollable = NULL;
+}
+
+int pg_graph_explore_ptr(struct pg_graph *graph,
+			 struct pg_brick *brick,
+			 struct pg_error **error)
+{
+	if (!brick) {
+		brick = get_any_brick(graph);
+		if (!brick) {
+			*error = pg_error_new("Graph %s is empty", graph->name);
+			return -1;
+		}
+	}
+
+	/* first, clear all graph */
+	empty_graph(graph);
+
+	/* explore new graph from brick */
+	GList *todo = NULL;
+	GList *done = NULL;
+
+	todo = g_list_prepend(todo, brick);
+	while (todo != NULL) {
```

Comment:
why a while ?
You initialize todo to NULL, after you prepend  brick to todo.
So you know that todo will contain only one elem.